### PR TITLE
CORE-12804: Propogate the error from the persistante service

### PR DESCRIPTION
**Pull request by yift-r3**
This should propagate the error type from the DB worker to the persistence client and support the `INVALID_ENTITY_UPDATE` <-> `InvalidEntityUpdateException` mapping.

The API pull request is in https://github.com/corda/corda-api/pull/1076

### CODE REVIEW

File: libs/utilities/src/main/kotlin/net/corda/utilities/Either.kt [hunk 1]
```diff
@@ -2,9 +2,16 @@ package net.corda.utilities
...
     data class Left<T>(
-        val a: T
+        val a: T,
     ) : Either<T, Nothing>
     data class Right<S>(
-        val b: S
+        val b: S,
     ) : Either<Nothing, S>
+
+    fun <U> mapRight(mapper: (S) -> U): Either<T, U> {
```
Comment (williamvigorr3) [inline libs/utilities/src/main/kotlin/net/corda/utilities/Either.kt:11]: Nit: I think `mapIfRight` would be slightly clearer?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceRPCProcessor.kt [hunk 2]
```diff
@@ -35,7 +37,11 @@ internal class MembershipPersistenceRPCProcessor(
         } catch (e: Exception) {
             val error = "Exception thrown while processing membership persistence request: ${e.message}"
             logger.warn(error)
-            PersistenceFailedResponse(error)
+            val kind = when (e) {
+                is InvalidEntityUpdateException -> ErrorKind.INVALID_ENTITY_UPDATE
```
Comment (williamvigorr3) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/MembershipPersistenceRPCProcessor.kt:41]: So I think a `PessimisticLockException` could also be thrown by the `ActivateMemberHandler`/`SuspendMemberHandler` and if that happens we want to send the same response back to the REST worker. For example that could happen if re-registration happened at the same time as suspension.

This is different to registration, we want the user to retry the REST request in this case.

**Review by williamvigorr3 — COMMENTED** [review]
I left a couple of comments. Maybe this is just my personal opinion but I would favour us moving away from using exceptions for error handling (we shouldn't do that now though). As they can propagate through multiple layers of the stack and it makes the code hard to reason about (probably at the expensive of more lines of code).

File: libs/utilities/src/main/kotlin/net/corda/utilities/Either.kt (hunk 1 above)
Comment (yift-r3) [reply libs/utilities/src/main/kotlin/net/corda/utilities/Either.kt:11]: Arrow has [`map`](https://old.arrow-kt.io/docs/apidocs/arrow-core/arrow.core/-either/map.html) and [`mapLeft`](https://old.arrow-kt.io/docs/apidocs/arrow-core/arrow.core/-either/map-left.html). Scala [`map`](https://www.scala-lang.org/api/2.13.6/scala/util/Either.html#map[B1](f:B=%3EB1):scala.util.Either[A,B1]) is also right only. Rust has [`map_laft`](https://docs.rs/either/latest/either/enum.Either.html#method.map_left) and [`map_right`](https://docs.rs/either/latest/either/enum.Either.html#method.map_right). I don't see a `mapIf...` is used in Either monad.  Maybe it should be `mapLeft` and a simple `map` to map the right side?

Comment (williamvigorr3) [reply libs/utilities/src/main/kotlin/net/corda/utilities/Either.kt:11]: I don't mind (do what you think is best).

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspensionActivationEntityOperations.kt [hunk 3]
```diff
@@ -69,21 +70,27 @@ internal class SuspensionActivationEntityOperations(
...
-                newStatus,
-                now,
-                currentMemberInfo.memberContext,
-                currentMemberInfo.memberSignatureKey,
-                currentMemberInfo.memberSignatureContent,
-                currentMemberInfo.memberSignatureSpec,
-                serializedMgmContext,
-                updatedSerial
+        try {
+            em.merge(
```
Comment (williamvigorr3) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspensionActivationEntityOperations.kt]: According to the EntityManager Java doc only `em.find` can throw a `PessimisticLockException`. I also think we should do the same thing if the GroupParameters conflict (although only inside `ActivateMemberHandler` and `SuspendMemberHandler`).

Comment (yift-r3) [reply components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspensionActivationEntityOperations.kt]: Yes, sorry, that is a lot more reasonable.

Comment (williamvigorr3) [reply components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspensionActivationEntityOperations.kt]: I think you missed the `GroupParameters` part out.

**Review by williamvigorr3 — DISMISSED** [review]
LGTM

**Review by dimosr — APPROVED** [review]

**Pull request merged**